### DISCUSSION TITLE
Consider supporting typed-strings

### DISCUSSION THREAD

**Discussion by springcomp** [jmespath-community/jmespath.spec#154] [Ideas]
In light of a future [extension of comparison](https://github.com/jmespath-community/jmespath.spec/discussions/91) to strings, it could be beneficial to support typed-strings and well-defined comparison semantics.

For instance comparing dates and times.
Comparing units.
Comparing guids.
Etc.

**Comment by springcomp** [comment]
Maybe extend `raw-string` to support a prefix.

- `` date'2023-03-03' ``
- `` uuid'2e71d99c-d01e-4695-adff-1db9997bd4a8' ``

That's where type coercion might be involved to promote type-inference. For instance, future date/time functions might be defined to operate on such strings and the prefix would not be needed in most cases.

Although one might be weary of implicit type conversions.

**Reply by gibson042** [reply]
If functionality like this gets added, I'd much rather see it take the form of new functions rather than new literal syntax, for reasons including backwards compatibility, extensibility, and minimizing complexity. So the examples above would look like `date('2023-03-03')` and `uuid('2e71d99c-d01e-4695-adff-1db9997bd4a8')`, and could probably be written today with no changes at all.

**Reply by kit494way** [reply]
> future date/time functions might be defined to operate on such strings and the prefix would not be needed in most cases.

I don't know in what cases the prefix is needed after date/time functions are implemented.
I think it is better to implement a `date()` function that parses a string and returns a unix timestamp, an ISO 8601 format string, or a string in the specified format.
